### ISSUE TITLE
0.7.0 regression: incorrect diff created for aws_security_group_rule with source group = target group

### ISSUE THREAD

**Issue by glasser** [hashicorp/terraform#7670]
This bug occurs on current HEAD and was introduced by @catsby in #7164 two weeks ago.  It is a regression from 0.6.

TF file:

```
resource "aws_security_group" "foo" {
  name_prefix = "tf-foo"
}

resource "aws_security_group_rule" "allow_self" {
  type = "ingress"
  from_port = 0
  to_port = 0
  protocol = "-1"
  security_group_id = "${aws_security_group.foo.id}"
  source_security_group_id = "${aws_security_group.foo.id}"
}
```

Run initial apply to create the resources. Note that it calculates self = false.

```
$ AWS_PROFILE=dev AWS_DEFAULT_REGION=us-east-1 terraform apply
aws_security_group.foo: Creating...
  description: "" => "Managed by Terraform"
  egress.#:    "" => "<computed>"
  ingress.#:   "" => "<computed>"
  name:        "" => "<computed>"
  name_prefix: "" => "tf-foo"
  owner_id:    "" => "<computed>"
  vpc_id:      "" => "<computed>"
aws_security_group.foo: Creation complete
aws_security_group_rule.allow_self: Creating...
  from_port:                "" => "0"
  protocol:                 "" => "-1"
  security_group_id:        "" => "sg-f0d8e58b"
  self:                     "" => "false"
  source_security_group_id: "" => "sg-f0d8e58b"
  to_port:                  "" => "0"
  type:                     "" => "ingress"
aws_security_group_rule.allow_self: Creation complete

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate
```

Now run a plan:

```
$ AWS_PROFILE=dev AWS_DEFAULT_REGION=us-east-1 terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

aws_security_group.foo: Refreshing state... (ID: sg-f0d8e58b)
aws_security_group_rule.allow_self: Refreshing state... (ID: sgrule-919438611)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

-/+ aws_security_group_rule.allow_self
    from_port:                "0" => "0"
    protocol:                 "-1" => "-1"
    security_group_id:        "sg-f0d8e58b" => "sg-f0d8e58b"
    self:                     "true" => "false" (forces new resource)
    source_security_group_id: "sg-f0d8e58b" => "sg-f0d8e58b"
    to_port:                  "0" => "0"
    type:                     "ingress" => "ingress"


Plan: 1 to add, 0 to change, 1 to destroy.
```

It wants to destroy and create the rule unnecessarily, because the new Read code for the resource figures out that self=true, but the Create code didn't figure that out.

Applying will destroy and create the rule, but still leave self=false in the state, so the problem continues to exist.

I've heard reports from teammates that in our actual setup it also triggers

```
* aws_security_group_rule.proxy_allow_proxy: [WARN] A duplicate Security Group rule was found on (sg-92d768e9).
```

although this does not seem to happen with this minimized reproduction.

**Comment by glasser** [comment]
I tried to fix this with https://github.com/meteor/terraform/commit/979e694f019c61d256ae01267694a580bbff34ca (which includes a test). However, it doesn't work; the test fails with:

```
	testing.go:264: Step 0 error: After applying this step, the plan was not empty:
		
		DIFF:
		
		DESTROY/CREATE: aws_security_group_rule.self_ingress
		  from_port:                "80" => "80"
		  protocol:                 "tcp" => "tcp"
		  security_group_id:        "sg-0adee371" => "sg-0adee371"
		  self:                     "true" => "false" (forces new resource)
		  source_security_group_id: "sg-0adee371" => "sg-0adee371"
		  to_port:                  "80" => "80"
		  type:                     "ingress" => "ingress"
```

I thought maybe I should set Computed:true on "self", but that doesn't work because you can't set Computed on anything that is the target of a ConflictsWith.

**Comment by modax** [comment]
I was hit by this too. Unfortunately, this pretty much makes 0.7.0-rc3 unusable in my real world AWS scenarios (unless I move away from source_security_group_id when it is equal to security_group_id and use self explicitly in that case which is a bit problematic for me). 

I wonder, once this is fixed, could you release -rc4? -rc3 has gotten many fixes since -rc2 and it would be nice to be able to test & use it for a larger scale stuff.

Regarding Computed and ConflictsWith validation, I personally think it is a bit excessive and probably could just be dropped as practically it only applies to tf code where semantically "Computed" is not a valid condition. Once this is done, "self" could be set as "Computed" with no "Default" and that might work (though I have not tested it).

**Comment by catsby** [comment]
Thanks for the detailed report! I've opened https://github.com/hashicorp/terraform/pull/7706 that should address this, can you check it out?

**Comment by glasser** [comment]
This looks good — hard to test right now as AWS is having problems, but given that's essentially a revert of the code that caused the problem, should be good!

**Comment by modax** [comment]
Hi @catsby,

I have just verified #7706 on my real world AWS code which previously had problems with rc3. Seems to work now (plan reports no changes in the infrastructure).

Thanks

**Comment by catsby** [comment]
I've merged #7706, thanks all!
